syntax = "proto3";

package bench;

import "google/protobuf/timestamp.proto";

service Bench {
  rpc GetGroupChat(GetGroupChatReq) returns (GroupChat);
}

message MessageContentText {
  string text = 1;  // plain text
}

message MessageContentChatCreated {}

message MessageContentChatEdited {}

message MessageContentUserInvited {
  uint64 target_user_id = 1;
}

message MessageContentUserLeft {}

message MessageContentUserMadeAdmin {
  uint64 target_user_id = 1;
}

message MessageContentUserRemovedAdmin {
  uint64 target_user_id = 1;
}

enum HostRequestStatus {
  HOST_REQUEST_STATUS_PENDING = 0;
  HOST_REQUEST_STATUS_ACCEPTED = 1;
  HOST_REQUEST_STATUS_REJECTED = 2;
  HOST_REQUEST_STATUS_CONFIRMED = 3;
  HOST_REQUEST_STATUS_CANCELLED = 4;
}

message MessageContentHostRequestStatusChanged {
  //this is the new host request status
  HostRequestStatus status = 1;
}

message Message {
  uint64 message_id = 1;
  uint64 author_user_id = 2;
  google.protobuf.Timestamp time = 3;
  oneof content {
    MessageContentText text = 4;  // plain text

    MessageContentChatCreated chat_created = 5;
    MessageContentChatEdited chat_edited = 6;
    MessageContentUserInvited user_invited = 7;
    MessageContentUserLeft user_left = 8;
    MessageContentUserMadeAdmin user_made_admin = 9;
    MessageContentUserRemovedAdmin user_removed_admin = 10;
    MessageContentHostRequestStatusChanged host_request_status_changed = 11;
  }
}

message GroupChat {
  uint64 group_chat_id = 1;
  string title = 2;
  repeated uint64 member_user_ids = 3; // inclusive of admins
  repeated uint64 admin_user_ids = 4;
  bool only_admins_invite = 5;
  bool is_dm = 6;
  google.protobuf.Timestamp created = 7;
  uint32 unseen_message_count = 8;
  uint64 last_seen_message_id = 9;
  Message latest_message = 10;
}

message GetGroupChatReq {
  uint64 group_chat_id = 1;
}
